receiver/opencensus/opencensus.go

```diff
@@ -173,18 +184,51 @@ func (ocr *Receiver) Stop() error {
 	var err = errAlreadyStopped
 	ocr.stopOnce.Do(func() {
 		// TODO: (@odeke-em) should we instead do (*grpc.Server).GracefulStop?
-		ocr.server.Stop()
+		ocr.serverGRPC.Stop()
+		_ = ocr.serverHTTP.Close()
 		_ = ocr.ln.Close()
 	})
 	return err
 }
 
-func (ocr *Receiver) startGRPCServer() error {
+func (ocr *Receiver) httpServer() *http.Server {
+	ocr.mu.Lock()
+	defer ocr.mu.Unlock()
+
+	if ocr.serverHTTP == nil {
+		ocr.serverHTTP = &http.Server{Handler: gatewayruntime.NewServeMux()}
+	}
+
+	return ocr.serverHTTP
+}
+
+func (ocr *Receiver) startServer() error {
 	err := errAlreadyStarted
 	ocr.startServerOnce.Do(func() {
 		errChan := make(chan error, 1)
 		go func() {
-			errChan <- ocr.server.Serve(ocr.ln)
+			// Register the grpc-gateway on the HTTP server mux
+			c := context.Background()
+			opts := []grpc.DialOption{grpc.WithInsecure()}
+			endpoint := ocr.ln.Addr().String()
+			mux := ocr.httpServer().Handler.(*gatewayruntime.ServeMux)
```

Comment:
is the gateway's http server going to make a recursive local connection to gRPC, or is it going to invoke gRPC handler directly without an extra network hop?

Reply:
It does make a recursive local connection to gRPC - not the most efficient, but to my knowledge the [grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway) project does not support that mode where you have it just call the gRPC service handler directly.

We could manually write a REST endpoint that converts the JSON to proto using [jsonpb](https://godoc.org/github.com/golang/protobuf/jsonpb) and then calls the service handler - or even one that just parses the JSON and executes the span receiver logic based on the JSON values. But either of those would require maintaining at least some extra code. My suggestion would be to hold off doing those optimizations for now, but I'm open if you think it should be done sooner.